File: swift/Sources/Fory/TypeInfo.swift
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Foundation

@inline(__always)
func normalizeRegisteredTypeID(_ typeID: TypeId) -> TypeId {
    switch typeID {
    case .namedEnum:
        return .enumType
    case .compatibleStruct, .namedCompatibleStruct, .namedStruct:
        return .structType
    case .namedExt:
        return .ext
    case .namedUnion, .union:
        return .typedUnion
    default:
        return typeID
    }
}

@inline(__always)
func namedRegisteredTypeID(for baseTypeID: TypeId, compatible: Bool) -> TypeId {
    switch baseTypeID {
    case .structType:
        return compatible ? .namedCompatibleStruct : .namedStruct
    case .enumType:
        return .namedEnum
    case .ext:
        return .namedExt
    case .typedUnion:
        return .namedUnion
    default:
        return baseTypeID
    }
}

@inline(__always)
func idRegisteredTypeID(for baseTypeID: TypeId, compatible: Bool) -> TypeId {
    switch baseTypeID {
    case .structType:
        return compatible ? .compatibleStruct : .structType
    default:
        return baseTypeID
    }
}

@inline(__always)
func resolveRegisteredWireTypeID(
    declaredTypeID: TypeId,
    registerByName: Bool,
    compatible: Bool
) -> TypeId {
    let baseTypeID = normalizeRegisteredTypeID(declaredTypeID)
    if registerByName {
        return namedRegisteredTypeID(for: baseTypeID, compatible: compatible)
    }
    return idRegisteredTypeID(for: baseTypeID, compatible: compatible)
}

@inline(__always)
func isAllowedRegisteredWireTypeID(
    _ wireTypeID: TypeId,
    declaredTypeID: TypeId,
    registerByName: Bool,
    compatible: Bool
) -> Bool {
    let baseTypeID = normalizeRegisteredTypeID(declaredTypeID)
    let expected = resolveRegisteredWireTypeID(
        declaredTypeID: declaredTypeID,
        registerByName: registerByName,
        compatible: compatible
    )
    if wireTypeID == expected {
        return true
    }
    if baseTypeID == .structType, compatible {
        return wireTypeID == .compatibleStruct ||
            wireTypeID == .namedCompatibleStruct ||
            wireTypeID == .structType ||
            wireTypeID == .namedStruct
    }
    if baseTypeID == .typedUnion {
        return wireTypeID == .union || (registerByName && wireTypeID == .namedUnion)
    }
    return false
}

@inline(__always)
func registeredWireTypeNeedsUserTypeID(_ wireTypeID: TypeId) -> Bool {
    switch wireTypeID {
    case .enumType, .structType, .ext, .typedUnion, .union:
        return true
    default:
        return false
    }
}

@inline(__always)
private func encodedTypeDefHeader(_ bytes: [UInt8]) throws -> UInt64 {
    guard bytes.count >= 8 else {
        throw ForyError.invalidData("encoded compatible type metadata must include an 8-byte header")
    }
    let buffer = ByteBuffer(bytes: bytes)
    return try buffer.readUInt64()
}

@inline(__always)
private func encodedTypeDefHeaderHash(_ bytes: [UInt8]) throws -> UInt64 {
    guard bytes.count >= 8 else {
        throw ForyError.invalidData("encoded compatible type metadata must include an 8-byte header")
    }
    let buffer = ByteBuffer(bytes: bytes)
    let header = try buffer.readUInt64()
    return header >> 14
}

private func fieldNeedsTypeInfo(_ fieldType: TypeMeta.FieldType) -> Bool {
    if let typeID = TypeId(rawValue: fieldType.typeID),
       TypeId.needsTypeInfoForField(typeID) {
        return true
    }
    return fieldType.generics.contains { fieldNeedsTypeInfo($0) }
}

private func encodedTypeDefHasUserTypeFields(_ fields: [TypeMeta.FieldInfo]) -> Bool {
    fields.contains { fieldNeedsTypeInfo($0.fieldType) }
}

public final class TypeInfo: @unchecked Sendable {
    static let uncached = TypeInfo(typeID: .unknown)

    let swiftTypeID: ObjectIdentifier
    let typeID: TypeId
    let userTypeID: UInt32?
    let registerByName: Bool
    let namespace: MetaString
    let typeName: MetaString
    let typeMeta: TypeMeta?
    public let compatibleTypeMeta: TypeMeta?
    let typeDefBytes: [UInt8]?
    let firstTypeDefBytes: [UInt8]?
    let typeDefHeader: UInt64?
    public let typeDefHeaderHash: UInt64?
    public let typeDefHasUserTypeFields: Bool

    private let reader: (ReadContext) throws -> Any
    private let compatibleReader: (ReadContext, TypeInfo) throws -> Any
    private let nativeWireTypeID: TypeId
    private let compatibleWireTypeID: TypeId

    init(
        swiftTypeID: ObjectIdentifier,
        typeID: TypeId,
        userTypeID: UInt32?,
        registerByName: Bool,
        namespace: MetaString,
        typeName: MetaString,
        typeMeta: TypeMeta? = nil,
        compatibleTypeMeta: TypeMeta? = nil,
        typeDefBytes: [UInt8]? = nil,
        firstTypeDefBytes: [UInt8]? = nil,
        typeDefHeader: UInt64? = nil,
        typeDefHeaderHash: UInt64? = nil,
        typeDefHasUserTypeFields: Bool = true,
        reader: @escaping (ReadContext) throws -> Any,
        compatibleReader: @escaping (ReadContext, TypeInfo) throws -> Any
    ) {
        self.swiftTypeID = swiftTypeID
        self.typeID = typeID
        self.userTypeID = userTypeID
        self.registerByName = registerByName
        self.namespace = namespace
        self.typeName = typeName
        self.typeMeta = typeMeta
        self.compatibleTypeMeta = compatibleTypeMeta ?? typeMeta
        self.typeDefBytes = typeDefBytes
        self.firstTypeDefBytes = firstTypeDefBytes
        self.typeDefHeader = typeDefHeader
        self.typeDefHeaderHash = typeDefHeaderHash
        self.typeDefHasUserTypeFields = typeDefHasUserTypeFields
        self.reader = reader
        self.compatibleReader = compatibleReader
        nativeWireTypeID = resolveRegisteredWireTypeID(
            declaredTypeID: typeID,
            registerByName: registerByName,
            compatible: false
        )
        compatibleWireTypeID = resolveRegisteredWireTypeID(
            declaredTypeID: typeID,
            registerByName: registerByName,
            compatible: true
        )
    }

    convenience init(
        swiftTypeID: ObjectIdentifier,
        typeID: TypeId,
        userTypeID: UInt32?,
        registerByName: Bool,
        namespace: MetaString,
        typeName: MetaString,
        fields: [TypeMeta.FieldInfo],
        reader: @escaping (ReadContext) throws -> Any,
        compatibleReader: @escaping (ReadContext, TypeInfo) throws -> Any
    ) throws {
        let compatibleWireTypeID = resolveRegisteredWireTypeID(
            declaredTypeID: typeID,
            registerByName: registerByName,
            compatible: true
        )
        let typeMeta = try TypeMeta(
            typeID: registerByName ? nil : compatibleWireTypeID.rawValue,
            userTypeID: registerByName ? nil : userTypeID,
            namespace: namespace,
            typeName: typeName,
            registerByName: registerByName,
            fields: fields,
            hasFieldsMeta: !fields.isEmpty
        )
        let typeDefBytes = try typeMeta.encode()
        var firstTypeDefBytes = [UInt8]()
        firstTypeDefBytes.reserveCapacity(typeDefBytes.count + 1)
        firstTypeDefBytes.append(0)
        firstTypeDefBytes.append(contentsOf: typeDefBytes)
        let typeDefHeader = try encodedTypeDefHeader(typeDefBytes)
        let typeDefHeaderHash = try encodedTypeDefHeaderHash(typeDefBytes)
        let canonicalTypeMeta = try TypeMeta(
            typeID: registerByName ? nil : compatibleWireTypeID.rawValue,
            userTypeID: registerByName ? nil : userTypeID,
            namespace: namespace,
            typeName: typeName,
            registerByName: registerByName,
            fields: fields,
            hasFieldsMeta: !fields.isEmpty,
            headerHash: typeDefHeaderHash
        )
        self.init(
            swiftTypeID: swiftTypeID,
            typeID: typeID,
            userTypeID: userTypeID,
            registerByName: registerByName,
            namespace: namespace,
            typeName: typeName,
            typeMeta: canonicalTypeMeta,
            compatibleTypeMeta: canonicalTypeMeta,
            typeDefBytes: typeDefBytes,
            firstTypeDefBytes: firstTypeDefBytes,
            typeDefHeader: typeDefHeader,
            typeDefHeaderHash: typeDefHeaderHash,
            typeDefHasUserTypeFields: encodedTypeDefHasUserTypeFields(fields),
            reader: reader,
            compatibleReader: compatibleReader
        )
    }

    convenience init(typeID: TypeId) {
        self.init(
            swiftTypeID: ObjectIdentifier(TypeInfo.self),
            typeID: typeID,
            userTypeID: nil,
            registerByName: false,
            namespace: MetaString.empty(specialChar1: ".", specialChar2: "_"),
            typeName: MetaString.empty(specialChar1: "$", specialChar2: "_"),
            reader: { _ in
                throw ForyError.invalidData("dynamic type \(typeID) uses runtime-only decode path")
            },
            compatibleReader: { _, _ in
                throw ForyError.invalidData("dynamic compatible type \(typeID) uses runtime-only decode path")
            }
        )
    }

    convenience init(dynamic typeInfo: TypeInfo, compatibleTypeMeta: TypeMeta) {
        self.init(
            swiftTypeID: typeInfo.swiftTypeID,
            typeID: typeInfo.typeID,
            userTypeID: typeInfo.userTypeID,
            registerByName: typeInfo.registerByName,
            namespace: typeInfo.namespace,
            typeName: typeInfo.typeName,
            typeMeta: typeInfo.typeMeta,
            compatibleTypeMeta: compatibleTypeMeta,
            typeDefBytes: typeInfo.typeDefBytes,
            firstTypeDefBytes: typeInfo.firstTypeDefBytes,
            typeDefHeader: typeInfo.typeDefHeader,
            typeDefHeaderHash: typeInfo.typeDefHeaderHash,
            typeDefHasUserTypeFields: typeInfo.typeDefHasUserTypeFields,
            reader: typeInfo.reader,
            compatibleReader: typeInfo.compatibleReader
        )
    }

    @inline(__always)
    func matches(
        typeID: TypeId,
        userTypeID: UInt32?,
        registerByName: Bool,
        namespace: String,
        typeName: String
    ) -> Bool {
        self.typeID == typeID &&
            self.userTypeID == userTypeID &&
            self.registerByName == registerByName &&
            self.namespace.value == namespace &&
            self.typeName.value == typeName
    }

    @inline(__always)
    func wireTypeID(compatible: Bool) -> TypeId {
        compatible ? compatibleWireTypeID : nativeWireTypeID
    }

    @inline(__always)
    func read(_ context: ReadContext, compatibleTypeInfo: TypeInfo? = nil) throws -> Any {
        if let compatibleTypeInfo {
            return try compatibleReader(context, compatibleTypeInfo)
        }
        if compatibleTypeMeta !== typeMeta {
            return try compatibleReader(context, self)
        }
        return try reader(context)
    }
}

File: csharp/src/Fory/TypeInfo.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System.Collections.Concurrent;
using System.Collections.Immutable;
using System.Reflection;

namespace Apache.Fory;

internal enum UserTypeKind
{
    Enum,
    Struct,
    Ext,
    TypedUnion,
}

public sealed class TypeInfo
{
    internal readonly record struct CompatibleTypeMetaCacheEntry(TypeMeta TypeMeta, byte[] EncodedBytes, ulong HeaderHash);

    private readonly object _serializer;
    private readonly Action<WriteContext, object?, bool> _writeDataObject;
    private readonly Func<ReadContext, object?> _readDataObject;
    private readonly Action<WriteContext, object?, RefMode, bool, bool> _writeObject;
    private readonly Func<ReadContext, RefMode, bool, object?> _readObject;
    private readonly Func<bool, IReadOnlyList<TypeMetaFieldInfo>> _compatibleTypeMetaFields;
    private static readonly IReadOnlyList<TypeMetaFieldInfo> EmptyTypeMetaFields = Array.Empty<TypeMetaFieldInfo>();

    private TypeInfo(
        Type type,
        object serializer,
        TypeId? builtInTypeId,
        UserTypeKind? userTypeKind,
        bool isDynamicType,
        bool isNullableType,
        bool isRefType,
        bool supportsCompatibleReadWithoutTypeMeta,
        object? defaultObject,
        bool evolving,
        bool isRegistered,
        uint? userTypeId,
        bool registerByName,
        MetaString? namespaceName,
        MetaString? typeName,
        Action<WriteContext, object?, bool> writeDataObject,
        Func<ReadContext, object?> readDataObject,
        Action<WriteContext, object?, RefMode, bool, bool> writeObject,
        Func<ReadContext, RefMode, bool, object?> readObject,
        Func<bool, IReadOnlyList<TypeMetaFieldInfo>> compatibleTypeMetaFields,
        TypeId? readWireTypeId,
        TypeMeta? readCompatibleTypeMeta)
    {
        Type = type;
        _serializer = serializer;
        BuiltInTypeId = builtInTypeId;
        UserTypeKind = userTypeKind;
        IsDynamicType = isDynamicType;
        IsNullableType = isNullableType;
        IsRefType = isRefType;
        SupportsCompatibleReadWithoutTypeMeta = supportsCompatibleReadWithoutTypeMeta;
        DefaultObject = defaultObject;
        Evolving = evolving;
        IsRegistered = isRegistered;
        UserTypeId = userTypeId;
        RegisterByName = registerByName;
        NamespaceName = namespaceName;
        TypeName = typeName;
        _writeDataObject = writeDataObject;
        _readDataObject = readDataObject;
        _writeObject = writeObject;
        _readObject = readObject;
        _compatibleTypeMetaFields = compatibleTypeMetaFields;
        ReadWireTypeId = readWireTypeId;
        ReadCompatibleTypeMeta = readCompatibleTypeMeta;
    }

    internal static TypeInfo Create<T>(Type type, Serializer<T> serializer)
    {
        Func<bool, IReadOnlyList<TypeMetaFieldInfo>> compatibleTypeMetaFields =
            CreateCompatibleTypeMetaFieldsProvider(serializer, out bool hasCompatibleTypeMetaFieldsProvider);
        (TypeId? builtInTypeId, UserTypeKind? userTypeKind, bool isDynamicType) = ResolveTypeShape(
            type,
            hasCompatibleTypeMetaFieldsProvider);
        bool evolving = ResolveStructEvolving(type, userTypeKind);
        bool isNullableType = !type.IsValueType || Nullable.GetUnderlyingType(type) is not null;
        bool isRefType = type != typeof(string) && !type.IsValueType;
        bool supportsCompatibleReadWithoutTypeMeta = ResolveSupportsCompatibleReadWithoutTypeMeta(serializer);
        return new TypeInfo(
            type,
            serializer,
            builtInTypeId,
            userTypeKind,
            isDynamicType,
            isNullableType,
            isRefType,
            supportsCompatibleReadWithoutTypeMeta,
            serializer.DefaultObject,
            evolving,
            isRegistered: false,
            userTypeId: null,
            registerByName: false,
            namespaceName: null,
            typeName: null,
            (context, value, hasGenerics) => WriteDataObject(serializer, context, value, hasGenerics),
            context => serializer.ReadData(context),
            (context, value, refMode, writeTypeInfo, hasGenerics) =>
                WriteObject(serializer, context, value, refMode, writeTypeInfo, hasGenerics),
            (context, refMode, readTypeInfo) => serializer.Read(context, refMode, readTypeInfo),
            compatibleTypeMetaFields,
            builtInTypeId,
            null);
    }

    private static Func<bool, IReadOnlyList<TypeMetaFieldInfo>> CreateCompatibleTypeMetaFieldsProvider(
        object serializer,
        out bool hasProvider)
    {
        MethodInfo? method = serializer.GetType().GetMethod(
            "CompatibleTypeMetaFields",
            BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic,
            null,
            [typeof(bool)],
            null);
        if (method is null || method.ReturnType != typeof(IReadOnlyList<TypeMetaFieldInfo>))
        {
            hasProvider = false;
            return EmptyCompatibleTypeMetaFields;
        }

        try
        {
            Delegate del = method.CreateDelegate(typeof(Func<bool, IReadOnlyList<TypeMetaFieldInfo>>), serializer);
            hasProvider = true;
            return (Func<bool, IReadOnlyList<TypeMetaFieldInfo>>)del;
        }
        catch
        {
            hasProvider = false;
            return EmptyCompatibleTypeMetaFields;
        }
    }

    private static IReadOnlyList<TypeMetaFieldInfo> EmptyCompatibleTypeMetaFields(bool _)
    {
        return EmptyTypeMetaFields;
    }

    private static bool ResolveSupportsCompatibleReadWithoutTypeMeta(object serializer)
    {
        MethodInfo? method = serializer.GetType().GetMethod(
            "SupportsCompatibleReadWithoutTypeMeta",
            BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic,
            null,
            Type.EmptyTypes,
            null);
        if (method is null || method.ReturnType != typeof(bool))
        {
            return false;
        }

        try
        {
            return (bool)method.Invoke(serializer, null)!;
        }
        catch
        {
            return false;
        }
    }

    private static bool ResolveStructEvolving(Type type, UserTypeKind? userTypeKind)
    {
        if (userTypeKind != Apache.Fory.UserTypeKind.Struct)
        {
            return true;
        }

        Type structType = Nullable.GetUnderlyingType(type) ?? type;
        ForyObjectAttribute? attribute = structType.GetCustomAttribute<ForyObjectAttribute>();
        return attribute?.Evolving ?? true;
    }

    private static void WriteDataObject<T>(Serializer<T> serializer, WriteContext context, object? value, bool hasGenerics)
    {
        serializer.WriteData(context, CoerceRuntimeValue(serializer, value), hasGenerics);
    }

    private static void WriteObject<T>(
        Serializer<T> serializer,
        WriteContext context,
        object? value,
        RefMode refMode,
        bool writeTypeInfo,
        bool hasGenerics)
    {
        serializer.Write(context, CoerceRuntimeValue(serializer, value), refMode, writeTypeInfo, hasGenerics);
    }

    private static T CoerceRuntimeValue<T>(Serializer<T> serializer, object? value)
    {
        if (value is T typed)
        {
            return typed;
        }

        if (value is null && default(T) is null)
        {
            return serializer.DefaultValue;
        }

        throw new InvalidDataException(
            $"serializer {serializer.GetType().Name} expected value of type {typeof(T)}, got {value?.GetType()}");
    }

    private static (TypeId? BuiltInTypeId, UserTypeKind? UserTypeKind, bool IsDynamicType) ResolveTypeShape(
        Type type,
        bool hasCompatibleTypeMetaFieldsProvider)
    {
        Type? nullableType = Nullable.GetUnderlyingType(type);
        if (nullableType is not null)
        {
            return ResolveTypeShape(nullableType, hasCompatibleTypeMetaFieldsProvider);
        }

        if (TryResolveBuiltInTypeId(type, out TypeId builtInTypeId))
        {
            return (builtInTypeId, null, false);
        }

        if (type == typeof(object))
        {
            return (null, null, true);
        }

        if (type.IsEnum)
        {
            return (null, Apache.Fory.UserTypeKind.Enum, false);
        }

        if (typeof(Union).IsAssignableFrom(type))
        {
            return (null, Apache.Fory.UserTypeKind.TypedUnion, false);
        }

        if (hasCompatibleTypeMetaFieldsProvider)
        {
            return (null, Apache.Fory.UserTypeKind.Struct, false);
        }

        return (null, Apache.Fory.UserTypeKind.Ext, false);
    }

    private static bool TryResolveBuiltInTypeId(Type type, out TypeId typeId)
    {
        if (type == typeof(bool))
        {
            typeId = TypeId.Bool;
            return true;
        }

        if (type == typeof(sbyte))
        {
            typeId = TypeId.Int8;
            return true;
        }

        if (type == typeof(short))
        {
            typeId = TypeId.Int16;
            return true;
        }

        if (type == typeof(int))
        {
            typeId = TypeId.VarInt32;
            return true;
        }

        if (type == typeof(long))
        {
            typeId = TypeId.VarInt64;
            return true;
        }

        if (type == typeof(byte))
        {
            typeId = TypeId.UInt8;
            return true;
        }

        if (type == typeof(ushort))
        {
            typeId = TypeId.UInt16;
            return true;
        }

        if (type == typeof(uint))
        {
            typeId = TypeId.VarUInt32;
            return true;
        }

        if (type == typeof(ulong))
        {
            typeId = TypeId.VarUInt64;
            return true;
        }

        if (type == typeof(float))
        {
            typeId = TypeId.Float32;
            return true;
        }

        if (type == typeof(double))
        {
            typeId = TypeId.Float64;
            return true;
        }

        if (type == typeof(string))
        {
            typeId = TypeId.String;
            return true;
        }

        if (type == typeof(byte[]))
        {
            typeId = TypeId.Binary;
            return true;
        }

        if (type == typeof(bool[]))
        {
            typeId = TypeId.BoolArray;
            return true;
        }

        if (type == typeof(sbyte[]))
        {
            typeId = TypeId.Int8Array;
            return true;
        }

        if (type == typeof(short[]))
        {
            typeId = TypeId.Int16Array;
            return true;
        }

        if (type == typeof(int[]))
        {
            typeId = TypeId.Int32Array;
            return true;
        }

        if (type == typeof(long[]))
        {
            typeId = TypeId.Int64Array;
            return true;
        }

        if (type == typeof(ushort[]))
        {
            typeId = TypeId.UInt16Array;
            return true;
        }

        if (type == typeof(uint[]))
        {
            typeId = TypeId.UInt32Array;
            return true;
        }

        if (type == typeof(ulong[]))
        {
            typeId = TypeId.UInt64Array;
            return true;
        }

        if (type == typeof(float[]))
        {
            typeId = TypeId.Float32Array;
            return true;
        }

        if (type == typeof(double[]))
        {
            typeId = TypeId.Float64Array;
            return true;
        }

        if (type == typeof(DateOnly))
        {
            typeId = TypeId.Date;
            return true;
        }

        if (type == typeof(DateTimeOffset) || type == typeof(DateTime))
        {
            typeId = TypeId.Timestamp;
            return true;
        }

        if (type == typeof(TimeSpan))
        {
            typeId = TypeId.Duration;
            return true;
        }

        if (type.IsArray)
        {
            typeId = TypeId.List;
            return true;
        }

        if (type.IsGenericType)
        {
            Type genericType = type.GetGenericTypeDefinition();
            if (genericType == typeof(List<>) ||
                genericType == typeof(LinkedList<>) ||
                genericType == typeof(Queue<>) ||
                genericType == typeof(Stack<>))
            {
                typeId = TypeId.List;
                return true;
            }

            if (genericType == typeof(HashSet<>) ||
                genericType == typeof(SortedSet<>) ||
                genericType == typeof(ImmutableHashSet<>))
            {
                typeId = TypeId.Set;
                return true;
            }

            if (genericType == typeof(Dictionary<,>) ||
                genericType == typeof(SortedDictionary<,>) ||
                genericType == typeof(SortedList<,>) ||
                genericType == typeof(ConcurrentDictionary<,>) ||
                genericType == typeof(NullableKeyDictionary<,>))
            {
                typeId = TypeId.Map;
                return true;
            }

            if (genericType == typeof(Nullable<>))
            {
                Type? underlying = Nullable.GetUnderlyingType(type);
                if (underlying is not null)
                {
                    return TryResolveBuiltInTypeId(underlying, out typeId);
                }
            }
        }

        typeId = default;
        return false;
    }

    public Type Type { get; }

    public bool IsBuiltinType => BuiltInTypeId.HasValue;

    public TypeId? BuiltInTypeId { get; }

    public bool IsUserType => UserTypeKind.HasValue;

    internal UserTypeKind? UserTypeKind { get; }

    public bool IsDynamicType { get; }

    public bool IsNullableType { get; }

    public bool IsRefType { get; }

    public bool SupportsCompatibleReadWithoutTypeMeta { get; }

    public object? DefaultObject { get; }

    internal bool Evolving { get; }

    internal TypeId? ReadWireTypeId { get; }

    internal TypeMeta? ReadCompatibleTypeMeta { get; }

    internal Type SerializerType => _serializer.GetType();

    public bool NeedsTypeInfoForField()
    {
        if (IsDynamicType)
        {
            return true;
        }

        if (!UserTypeKind.HasValue)
        {
            return false;
        }

        return UserTypeKind.Value is Apache.Fory.UserTypeKind.Struct or Apache.Fory.UserTypeKind.Ext;
    }

    internal Serializer<T> RequireSerializer<T>()
    {
        if (_serializer is Serializer<T> serializer)
        {
            return serializer;
        }

        throw new InvalidDataException($"serializer type mismatch for {typeof(T)}");
    }

    internal void WriteDataObject(WriteContext context, object? value, bool hasGenerics)
    {
        _writeDataObject(context, value, hasGenerics);
    }

    internal object? ReadDataObject(ReadContext context)
    {
        return _readDataObject(context);
    }

    internal void WriteObject(WriteContext context, object? value, RefMode refMode, bool writeTypeInfo, bool hasGenerics)
    {
        _writeObject(context, value, refMode, writeTypeInfo, hasGenerics);
    }

    internal object? ReadObject(ReadContext context, RefMode refMode, bool readTypeInfo)
    {
        return _readObject(context, refMode, readTypeInfo);
    }

    internal IReadOnlyList<TypeMetaFieldInfo> CompatibleTypeMetaFields(bool trackRef)
    {
        return _compatibleTypeMetaFields(trackRef);
    }

    internal bool IsRegistered { get; }

    internal uint? UserTypeId { get; }

    internal bool RegisterByName { get; }

    internal MetaString? NamespaceName { get; }

    internal MetaString? TypeName { get; }

    internal TypeInfo WithTypeIdRegistration(uint userTypeId)
    {
        return new TypeInfo(
            Type,
            _serializer,
            BuiltInTypeId,
            UserTypeKind,
            IsDynamicType,
            IsNullableType,
            IsRefType,
            SupportsCompatibleReadWithoutTypeMeta,
            DefaultObject,
            Evolving,
            isRegistered: true,
            userTypeId: userTypeId,
            registerByName: false,
            namespaceName: null,
            typeName: null,
            _writeDataObject,
            _readDataObject,
            _writeObject,
            _readObject,
            _compatibleTypeMetaFields,
            ReadWireTypeId,
            ReadCompatibleTypeMeta);
    }

    internal TypeInfo WithTypeNameRegistration(MetaString namespaceName, MetaString typeName)
    {
        return new TypeInfo(
            Type,
            _serializer,
            BuiltInTypeId,
            UserTypeKind,
            IsDynamicType,
            IsNullableType,
            IsRefType,
            SupportsCompatibleReadWithoutTypeMeta,
            DefaultObject,
            Evolving,
            isRegistered: true,
            userTypeId: null,
            registerByName: true,
            namespaceName: namespaceName,
            typeName: typeName,
            _writeDataObject,
            _readDataObject,
            _writeObject,
            _readObject,
            _compatibleTypeMetaFields,
            ReadWireTypeId,
            ReadCompatibleTypeMeta);
    }

    internal TypeInfo WithRegistrationFrom(TypeInfo source)
    {
        if (!source.IsRegistered)
        {
            return this;
        }

        if (source.RegisterByName)
        {
            if (!source.NamespaceName.HasValue || !source.TypeName.HasValue)
            {
                throw new InvalidDataException("missing type name metadata for name-registered type");
            }

            return WithTypeNameRegistration(source.NamespaceName.Value, source.TypeName.Value);
        }

        if (!source.UserTypeId.HasValue)
        {
            throw new InvalidDataException("missing user type id metadata for id-registered type");
        }

        return WithTypeIdRegistration(source.UserTypeId.Value);
    }

    internal TypeInfo WithReadTypeInfo(TypeId wireTypeId, TypeMeta? compatibleTypeMeta = null)
    {
        return new TypeInfo(
            Type,
            _serializer,
            BuiltInTypeId,
            UserTypeKind,
            IsDynamicType,
            IsNullableType,
            IsRefType,
            SupportsCompatibleReadWithoutTypeMeta,
            DefaultObject,
            Evolving,
            IsRegistered,
            UserTypeId,
            RegisterByName,
            NamespaceName,
            TypeName,
            _writeDataObject,
            _readDataObject,
            _writeObject,
            _readObject,
            _compatibleTypeMetaFields,
            wireTypeId,
            compatibleTypeMeta);
    }
}
